### ISSUE TITLE
PYTHON (KERAS) - How do I modify KERAS builtin “get batch” callback function?

### ISSUE THREAD

**Issue by manuelblancovalentin** [keras-team/keras#10056]
Hi, everyone.

I am currently working on an image classification problem for my PhD Thesis on oil/gas data processing. The data I need to classify is a SINGLE image from a borehole (like this one here, on the left: http://hager-richter.net/wp-content/uploads/2012/03/bore_graf.jpg), from which I have a certain lithological classification curve.

This means that, on one side I have a continuous image (like the one on the link), and on the other side I have a discrete curve with categorical numbers (say, 0, 1, 2, 3..., where each number indicates a certain lithological class, something like the type of rock that line of the image can be classified as). Let's call the input data (image) "X" and the classification curve "Y".

X has a shape of: (img_Height, img_Width, number_img_channels)
Y has a shape of: (img_Height, 1)

The idea is to use a CNN (raghakot resnet implementation: https://github.com/raghakot/keras-resnet) to create a model that petrophysicists can use to classify other borehole images.

Now, if I'm not mistaken, any image data inputed into a keras model should have the shape (number_of_samples, img_Height, img_Width, number_img_channels), so in order for this to work I need to modify the keras built-in batch function so that instead of getting a batch by taking something like:

`Xbatch = X[batch_index,:,:,:]`

it actually uses a certain window around each "batch_index", so that the function returns something like:

```
H = 21 #window height
Xbatch = np.zeros((len(batch_index)-1, img_Height, img_Width, number_img_channels)
for i in np.arange(0,len(batch_index)-1):
   Xbatch[i,:,:,:] = X[batch_index[i]+H, :, :].reshape((1, H, img_Width, number_img_channels))

```
The problem is that I have not idea which is the function (or file) of the keras module that I should modify in order to make this work. I think it might have something to do with the "keras.preprocessing.image.ImageDataGenerator" function, but I have not been able to find the right place where I should put my code.

This is the part of my code that initializes the model and trains it, using data augmentation (copy/pasted from raghakot implementation):

```
# Model definition
        model = resnet.ResnetBuilder.build_resnet_18((img_channels, __M__, __N__), nb_classes)
        model.compile(loss='categorical_crossentropy',
                      optimizer='adam',
                      metrics=['accuracy'])

        if not data_augmentation:
            print('Not using data augmentation.')
            model.fit(XTrain, YTrain,
                      batch_size=batch_size,
                      nb_epoch=nb_epoch,
                      validation_data=(XTest, YTest),
                      shuffle=True,
                      callbacks=[lr_reducer, early_stopper, csv_logger])
        else:
            print('Using real-time data augmentation.')
            # This will do preprocessing and realtime data augmentation:
            datagen = ImageDataGenerator(
                featurewise_center=False,  # set input mean to 0 over the dataset
                samplewise_center=False,  # set each sample mean to 0
                featurewise_std_normalization=False,  # divide inputs by std of the dataset
                samplewise_std_normalization=False,  # divide each input by its std
                zca_whitening=False,  # apply ZCA whitening
                rotation_range=0,  # randomly rotate images in the range (degrees, 0 to 180)
                width_shift_range=0.1,  # randomly shift images horizontally (fraction of total width)
                height_shift_range=0.1,  # randomly shift images vertically (fraction of total height)
                horizontal_flip=True,  # randomly flip images
                vertical_flip=False)  # randomly flip images

            # Compute quantities required for featurewise normalization
            # (std, mean, and principal components if ZCA whitening is applied).
            datagen.fit(XTrain)

            train_generator = datagen.flow(XTrain,YTrain,batch_size=batch_size)
            # Fit the model on the batches generated by datagen.flow().
            model.fit_generator(datagen.flow(XTrain, YTrain, batch_size=batch_size),
                                steps_per_epoch=XTrain.shape[0] // batch_size,
                                validation_data=(XTest, YTest),
                                epochs=nb_epoch, verbose=1, max_q_size=100,
                                callbacks=[lr_reducer, early_stopper, csv_logger])
```
Thanks in advance!

**Comment by tRosenflanz** [comment]
Looks like `_flow_index` in `ImageDataGenerator` is the function that handles the slicing of the arrays to batches  - you should probably look to modify it but I might be wrong

**Comment by manuelblancovalentin** [comment]
First of all, thanks for your comment @tRosenflanz!

As suggested, I tried to find out whether `_flow_index` was the function that iterated the batches on my code, so I simply added a print line at the beginning of it. However nothing appeared on the console when I called the function again so I tried to take a look a little bit further into it and found out that the function that is handling the batch iteration is the `NumpyArrayIterator`. I placed a `print('HEY!')` right after its `__init__` definition and it did print the message on the console. 

Thus, I am going to modify the `ImageDataGenerator` function and add a new class (say, `NumpySlidingIterator`) and see if I solve this problem. 

I'll post any news regarding this issue.

**Comment by manuelblancovalentin** [comment]
If anyone is interested in this, I was able to solve the problem by creating a copy of "keras.models", which I called "SlidingModels", and then adding five functions inside the Model class definition: `_fit_sliding_loop` (a modified copy of `_fit_loop function`), `_predict_sliding_loop` (a modified copy of `_predict_loop`), `_test_sliding_loop` (a modified version of `_test_loop`), `fit_sliding` (a modified version of `fit` function) and `predict_sliding` (a modified version of `predict` function).

A part from that I had to create the function `_slice_sliding_arrays` (modified copy of `_slice_arrays` function), outside the Model class definition.

Doing that I was able to train my CNN using a sliding window. The main stuff is on `_fit_sliding_loop`, inside the `for batch_index, (batch_start, batch_end) in enumerate(batches):`, where I added the sliding window cropping of the data. Now this loop looks like this:

`if shuffle == 'batch':
                    index_array = _batch_shuffle(index_array, batch_size)
                elif shuffle:
                    np.random.shuffle(index_array)

                batches = _make_batches(num_train_samples, batch_size)
                for batch_index, (batch_start, batch_end) in enumerate(batches):
                    batch_ids = index_array[batch_start:batch_end]
                    try:
                        if isinstance(ins[-1], float):
                            # Do not slice the training phase flag.
                            # slice x
                            start = batch_ids.tolist()
                            start = [training_indexes[s] for s in start]
                            xbatch = np.expand_dims([x[s:s+sliding_window,:] for s in start],axis=3)
                            ybatch = y[start,:]
                            ins_batch = (xbatch,ybatch,(ins[-2][batch_ids]),(ins[-1]))
                        else:
                            start = batch_ids.tolist()
                            start = [training_indexes[s] for s in start]
                            xbatch = np.expand_dims([x[s:s+sliding_window,:] for s in start],axis=3)
                            ybatch = y[start,:]
                            ins_batch = (xbatch,ybatch,(ins[-2][batch_ids]),(ins[-1]))
                    except TypeError:
                        raise TypeError('TypeError while preparing batch. '
                                        'If using HDF5 input data, '
                                        'pass shuffle="batch".')
                    batch_logs = {}
                    batch_logs['batch'] = batch_index
                    batch_logs['size'] = len(batch_ids)
                    callbacks.on_batch_begin(batch_index, batch_logs)
                    for i in indices_for_conversion_to_dense:
                        ins_batch[i] = ins_batch[i].toarray()

                    outs = f(ins_batch)
                    if not isinstance(outs, list):
                        outs = [outs]
                    for l, o in zip(out_labels, outs):
                        batch_logs[l] = o

                    callbacks.on_batch_end(batch_index, batch_logs)
                    if callback_model.stop_training:
                        break

                    if batch_index == len(batches) - 1:  # Last batch.
                        if do_validation:
                            val_outs = self._test_sliding_loop(x=x,y=y, validating_indexes=validating_indexes,val_f=val_f, val_ins=val_ins,
                                                       batch_size=batch_size, sliding_window = sliding_window,
                                                       verbose=0)
                            if not isinstance(val_outs, list):
                                val_outs = [val_outs]
                            # Same labels assumed.
                            for l, o in zip(out_labels, val_outs):
                                epoch_logs['val_' + l] = o`

Anyway, If anyone wants more information about the code, send me a message.
